File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Cluster-Mode/Perl/vollist.pl
#============================================================#
#                                                            #
#                                                            #
# vollist.pl                                                 #
#                                                            #
# Sample code to list the volumes available in the cluster.  #
#                                                            #
# This sample code is supported from Cluster-Mode            #
# Data ONTAP 8.1 onwards.                                    #
#                                                            #
# Copyright 2011 NetApp, Inc. All rights reserved.           #
# Specifications subject to change without notice.           #
#                                                            #
#============================================================#

use lib "../../../../../lib/perl/NetApp";
use NaServer;
use NaElement;
use strict;

my $args = $#ARGV + 1;
my ($server, $ipaddr, $user, $passwd);


sub print_usage_and_exit() {
    print("\nUsage: \n");
    print("vollist.pl <cluster/vserver> <user> <passwd> [-v <vserver-name>] \n");
    print("<cluster>             -- IP address of the cluster \n");
    print("<vserver>             -- IP address of the vserver \n");
    print("<user>                -- User name \n");
    print("<passwd>              -- Password \n");
    print("<vserver-name>        -- Name of the vserver \n\n");
    print("Note: \n");
    print(" -v switch is required when you want to tunnel the command to a vserver using cluster interface \n\n");
    exit (-1);
}

sub list_volumes {
    my ($in, $out, $tag);

    $tag = "";
   if ($args > 3) {
        if ($args < 5 || $ARGV[3] ne ("-v")) {
            print_usage_and_exit();
        }
        $server->set_vserver($ARGV[4]);
    }
    while (defined($tag)) {
        $in = NaElement->new("volume-get-iter");
        if ($tag ne "") {
            $in->child_add_string("tag", $tag);
        }
        my $out = $server->invoke_elem($in);
        if ($out->results_status() eq "failed") {
            print($out->results_reason() ."\n");
            exit(-1);
        }
        if ($out->child_get_int("num-records") == 0) {
            return;
        }
        $tag = $out->child_get_string("next-tag");
        my @volList = $out->child_get("attributes-list")->children_get();
        my $volInfo;
        my ($vserverName, $volName, $aggrName, $volType, $volState, $size, $availSize);
        print("----------------------------------------------------\n");
        foreach $volInfo (@volList) {
            $vserverName = $volName = $aggrName = $volType = $volState = $size = $availSize = "";
            my $volIdAttrs = $volInfo->child_get("volume-id-attributes");
            if ($volIdAttrs) {
                $vserverName = $volIdAttrs->child_get_string("owning-vserver-name");
                $volName = $volIdAttrs->child_get_string("name");
                $aggrName = $volIdAttrs->child_get_string("containing-aggregate-name");
                $volType = $volIdAttrs->child_get_string("type");
            }
            print("Vserver Name            : $vserverName \n");
            print("Volume Name             : $volName \n");
            print("Aggregate Name          : $aggrName \n");
            print("Volume type             : $volType \n");
            my $volStateAttrs = $volInfo->child_get("volume-state-attributes");
            if ($volStateAttrs) {
                $volState = $volStateAttrs->child_get_string("state");
            }    
            print("Volume state            : $volState \n");
            my $volSizeAttrs = $volInfo->child_get("volume-space-attributes");
            if ($volSizeAttrs) {
                $size = $volSizeAttrs->child_get_string("size");
                $availSize = $volSizeAttrs->child_get_string("size-available");
            }
            print("Size (bytes)            : $size \n");
            print("Available Size (bytes)  : $availSize \n");
            print("----------------------------------------------------\n");
        }
    }
}

sub main() {

    if ($args < 3) {
        print_usage_and_exit();
    }
    $ipaddr = $ARGV[0];
    $user = $ARGV[1];
    $passwd = $ARGV[2];

    $server = NaServer->new($ipaddr, 1, 15);
    $server->set_style("LOGIN");
    $server->set_admin_user($user, $passwd);
    $server->set_transport_type("HTTP");
    list_volumes();
}

main();

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Cluster-Mode/DotNet/VB.NET/vollist/vollist.vb
'============================================================'
'                                                            '
' vollist.vb                                                 '
'                                                            '
' Sample code to list the volumes available in the cluster.  '
'                                                            '
' This sample code is supported from Cluster-Mode            '
' Data ONTAP 8.1 onwards.                                    '
'                                                            '
' Copyright 2011 NetApp, Inc. All rights reserved.           '
' Specifications subject to change without notice.           '
'                                                            '
'============================================================'

Imports System
Imports System.Collections.Generic
Imports NetApp.Manage

Module VolList
    Dim server As NaServer
    Dim args As String()

    Public Sub PrintUsageAndExit()
        Console.WriteLine(Environment.NewLine & "Usage:")
        Console.WriteLine("vollist <cluster/vserver> <user> <passwd> [-v <vserver-name>]")
        Console.WriteLine("<cluster>             -- IP address of the cluster")
        Console.WriteLine("<vserver>             -- IP address of the vserver")
        Console.WriteLine("<user>                -- User name")
        Console.WriteLine("<passwd>              -- Password")
        Console.WriteLine("<vserver-name>        -- Name of the vserver")
        Console.WriteLine("Note: ")
        Console.WriteLine(" -v switch is required when you want to tunnel the command to a vserver using cluster interface")
        Environment.Exit(-1)
    End Sub

    Sub ListVolumes()
        Dim xi, xo As NaElement
        Dim tag As String = ""
        Dim flag As Boolean = True
        Dim vserverName, volName, aggrName, volType, volState, size, availSize As String
        While (flag = True)
            If (args.Length > 4) Then
                If (args.Length < 6 Or (Not args(4).Equals("-v"))) Then
                    PrintUsageAndExit()
                End If
                server.Vserver = args(5)
            End If
            xi = New NaElement("volume-get-iter")
            If (Not tag.Equals("")) Then
                xi.AddNewChild("tag", tag)
            End If
            xo = server.InvokeElem(xi)
            If (xo.GetChildIntValue("num-records", 0) = 0) Then
                Console.WriteLine("No volume(s) information available")
                Return
            End If
            tag = xo.GetChildContent("next-tag")
            If (tag = Nothing) Then
                flag = False
            End If
            Dim volList As List(Of NaElement) = xo.GetChildByName("attributes-list").GetChildren()
            Dim total As Integer = volList.Count
            Dim index As Integer
            Dim volInfo As NaElement
            Dim volStateAttrs As NaElement
            Dim volSizeAttrs As NaElement

            Console.WriteLine("----------------------------------------------------")
            For index = 0 To total - 1
                volInfo = volList(index)
                vserverName = ""
                volName = ""
                aggrName = ""
                volType = ""
                volState = ""
                size = ""
                availSize = ""
                Dim volIdAttrs As NaElement = volInfo.GetChildByName("volume-id-attributes")
                If (Not (volIdAttrs Is Nothing)) Then
                    vserverName = volIdAttrs.GetChildContent("owning-vserver-name")
                    volName = volIdAttrs.GetChildContent("name")
                    aggrName = volIdAttrs.GetChildContent("containing-aggregate-name")
                    volType = volIdAttrs.GetChildContent("type")
                End If

                Console.WriteLine("Vserver Name            : " + IIf(vserverName <> Nothing, vserverName, ""))
                Console.WriteLine("Volume Name             : " + IIf(volName <> Nothing, volName, ""))
                Console.WriteLine("Aggregate Name          : " + IIf(aggrName <> Nothing, aggrName, ""))
                Console.WriteLine("Volume type             : " + IIf(volType <> Nothing, volType, ""))
                volStateAttrs = volInfo.GetChildByName("volume-state-attributes")
                If (Not (volIdAttrs Is Nothing)) Then
                    volState = volStateAttrs.GetChildContent("state")
                End If
                Console.WriteLine("Volume state            : " + IIf(volState <> Nothing, volState, ""))
                volSizeAttrs = volInfo.GetChildByName("volume-space-attributes")
                If (Not (volSizeAttrs Is Nothing)) Then
                    size = volSizeAttrs.GetChildContent("size")
                    availSize = volSizeAttrs.GetChildContent("size-available")
                End If
                Console.WriteLine("Size (bytes)            : " + IIf(size <> Nothing, size, ""))
                Console.WriteLine("Available Size (bytes)  : " + IIf(availSize <> Nothing, availSize, ""))
                Console.WriteLine("----------------------------------------------------")
            Next index
        End While
    End Sub

    Sub Main()
        args = Environment.GetCommandLineArgs()

        If (args.Length < 4) Then
            PrintUsageAndExit()
            Exit Sub
        End If

        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)

        Try
            'Initialize connection to server, and request version 1.15 of the API set.
            server = New NaServer(serverName, 1, 15)

            server.ServerType = NaServer.SERVER_TYPE.FILER
            server.TransportType = NaServer.TRANSPORT_TYPE.HTTP
            server.SetAdminUser(user, passwd)

            ListVolumes()
           
        Catch ex As NaException
            Console.WriteLine(ex.Message)
        Catch ex As Exception
            Console.WriteLine(ex.Message)
        End Try
    End Sub

End Module
